preprocess.py

```diff
@@ -0,0 +1,290 @@
+        try:
+            # Convert to integer
+            n = int(num)
+            # Use num2words with Arabic language
+            arabic_word = num2words(n, lang='ar')
+            # Replace the number with its word form using word boundaries
+            text = re.sub(re.escape(num), arabic_word, text)
+        except (ValueError, NotImplementedError):
+            # Skip if conversion fails
+            continue
+    
+    return text
+
+def filter_non_arabic_words(text):
+    """Remove non-Arabic words from text."""
+    # Arabic Unicode range (includes Arabic, Persian, Urdu characters)
+    arabic_pattern = re.compile(r'[\u0600-\u06FF\u0750-\u077F\u08A0-\u08FF\u0660-\u0669]+')
+    
+    # Split text into words
+    words = text.split()
+    
+    # Keep only words that contain Arabic characters
+    arabic_words = []
+    for word in words:
+        # Check if the word ONLY contains Arabic characters
+        if arabic_pattern.search(word):
+            arabic_words.append(word)
+    
+    # Join the Arabic words back into text
+    return ' '.join(arabic_words)
+
+def phonemize(text, global_phonemizer, tokenizer):
+    """Convert text to phonemes and token IDs."""
+    text = convert_numbers_to_arabic_words(text)
+    text = filter_non_arabic_words(text)
+
+    text = remove_accents(text)
+    # TODO: English PL-BERT uses normalize_text, but Arabic PL-BERT does not and may need to use in the future. However, the code is highly manual and the Arabic version implementation TBD.
+    tokens = tokenizer.tokenize(text)
+
+    token_ids = [tokenizer.convert_tokens_to_ids(token) for token in tokens]
+    phonemes = [global_phonemizer.phonemize([token.replace("#", "")], strip=True)[0] if token not in string.punctuation else token for token in tokens]
+        
+    return {'token_ids' : token_ids, 'phonemes': phonemes}
+
+def parse_args():
+    """Parse command line arguments."""
+    parser = argparse.ArgumentParser(description="Preprocess text data for phoneme-level BERT training")
+    parser.add_argument("--config_path", type=str, default="external/pl_bert/Configs/config.yml", help="Path to config file")
+    parser.add_argument("--local_dataset_path", type=str, default=None, help="Path to local dataset (if using local data)")
+    parser.add_argument("--use_local_dataset", action="store_true", help="Use local dataset instead of downloading from HuggingFace")
+    return parser.parse_args()
+
+def process_shard(args: Tuple[int, str, Dataset, Any, Any, int]) -> Tuple[int, bool]:
+    """Process a single dataset shard."""
+    i, root_directory, dataset, global_phonemizer, tokenizer, num_shards = args
+    directory = os.path.join(root_directory, f"shard_{i}")
+    
+    if os.path.exists(directory):
+        print(f"Shard {i} already exists!")
+        return i, True
+    
+    try:
+        print(f'Processing shard {i} ...')
+        shard = dataset.shard(num_shards=num_shards, index=i)
+        processed_dataset = shard.map(
```

Comment:
btw you can use nproc to run the map function on multiple processor and maybe instead process shards sequentially which could be better in case the machine crashes or something. Just mentioning it, no need to change what you have, I think multiprocessing on the same shard doesn't scale as well as what you have cause it needs to split and combine.